File: api/routes/traces.py
"""
Query trace routes for frontend visualization.

    GET  /traces                -- list recent traces
    GET  /traces/{trace_id}    -- get a single trace with full phases
    DELETE /traces/{trace_id}  -- delete a trace
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel

from ..deps import get_state
from ..state import AppState

router = APIRouter(prefix="/api/v1/traces", tags=["traces"])


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------


class TraceSummary(BaseModel):
    trace_id: str
    query: str
    timestamp: Any
    total_ms: int
    total_llm_ms: int
    total_llm_calls: int
    answer_model: str | None = None
    finish_reason: str | None = None
    citations_used: list[str]


class TraceDetail(TraceSummary):
    answer_text: str | None = None
    trace_json: dict[str, Any]
    metadata_json: dict[str, Any]


from ..schemas import PaginatedResponse


class TraceListResponse(PaginatedResponse):
    pass


# ---------------------------------------------------------------------------
# Routes
# ---------------------------------------------------------------------------


@router.get("", response_model=TraceListResponse)
def list_traces(
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
    state: AppState = Depends(get_state),
) -> TraceListResponse:
    rows = state.store.list_traces(limit=limit, offset=offset)
    items = [
        TraceSummary(
            trace_id=r["trace_id"],
            query=r["query"],
            timestamp=r["timestamp"],
            total_ms=r["total_ms"],
            total_llm_ms=r["total_llm_ms"],
            total_llm_calls=r["total_llm_calls"],
            answer_model=r.get("answer_model"),
            finish_reason=r.get("finish_reason"),
            citations_used=r.get("citations_used", []),
        )
        for r in rows
    ]
    total = state.store.count_traces()
    return TraceListResponse(items=items, total=total, limit=limit, offset=offset)


@router.get("/{trace_id}", response_model=TraceDetail)
def get_trace(
    trace_id: str,
    state: AppState = Depends(get_state),
) -> TraceDetail:
    row = state.store.get_trace(trace_id)
    if not row:
        raise HTTPException(status_code=404, detail="trace not found")
    return TraceDetail(**row)


@router.delete("/{trace_id}")
def delete_trace(
    trace_id: str,
    state: AppState = Depends(get_state),
):
    row = state.store.get_trace(trace_id)
    if not row:
        raise HTTPException(status_code=404, detail="trace not found")
    state.store.delete_trace(trace_id)
    return {"deleted": trace_id}

File: web/src/api/traces.js
/**
 * Traces API — 查询审计日志 (可视化检索流程)
 *
 * GET    /api/v1/traces              查询历史列表
 * GET    /api/v1/traces/{id}         单个 trace 详情 (含完整 phases)
 * DELETE /api/v1/traces/{id}         删除
 *
 * trace_json 结构 (用于前端渲染 Gantt/瀑布图):
 * {
 *   query, timestamp, total_ms, total_llm_ms, total_llm_calls,
 *   phases: [
 *     { name, started_at_ms, duration_ms, inputs, outputs, llm_calls, details }
 *   ],
 *   generation: { model, latency_ms, usage, citations_used, answer_length }
 * }
 */

import { get, del } from './client'

/**
 * 查询历史列表 (分页,按时间倒序)
 * @param {object} [params]
 * @param {number} [params.limit=50]
 * @param {number} [params.offset=0]
 * @returns {Promise<{
 *   items: Array<{
 *     trace_id: string, query: string, timestamp: string,
 *     total_ms: number, total_llm_ms: number, total_llm_calls: number,
 *     answer_model: string|null, finish_reason: string|null,
 *     citations_used: string[]
 *   }>,
 *   total: number, limit: number, offset: number
 * }>}
 */
export const listTraces = (params = {}) =>
  get('/api/v1/traces', { limit: 50, offset: 0, ...params })

/**
 * 单个 trace 详情 (含完整 trace_json)
 * @param {string} traceId
 * @returns {Promise<{
 *   trace_id: string, query: string, timestamp: string,
 *   total_ms: number, total_llm_ms: number, total_llm_calls: number,
 *   answer_text: string|null, answer_model: string|null,
 *   finish_reason: string|null, citations_used: string[],
 *   trace_json: TraceJSON, metadata_json: object
 * }>}
 */
export const getTrace = (traceId) => get(`/api/v1/traces/${traceId}`)

/**
 * 删除 trace
 * @param {string} traceId
 * @returns {Promise<{ deleted: string }>}
 */
export const deleteTrace = (traceId) => del(`/api/v1/traces/${traceId}`)
